Cython/Build/Tests/TestCyCache.py

```diff
@@ -26,38 +33,50 @@ def fresh_cythonize(self, *args, **kwargs):
         kwargs.update(cache=self.cache_dir)
         fresh_cythonize(*args, **kwargs)
 
+    def relative_lines_from_file(self, path, line, start, end):
+        return relative_lines_from_file(
+            path, line, start, end, fallback=self.fallback_lines)
+
     def test_cycache_switch(self):
+        a_filename = 'a.pyx'
         content1 = 'value = 1\n'
         content2 = 'value = 2\n'
-        a_pyx = os.path.join(self.src_dir, 'a.pyx')
+        a_pyx = os.path.join(self.src_dir, a_filename)
         a_c = a_pyx[:-4] + '.c'
+        
+        module_line_1 = LINE_BEFORE_IMPLEMENTATION.format(
+            filename=a_filename, at_line=1)
+        definition_1 = "PyDict_SetItem(__pyx_d, __pyx_n_s_value, __pyx_int_1)"
+        definition_2 = "PyDict_SetItem(__pyx_d, __pyx_n_s_value, __pyx_int_2)"
 
         write_file(a_pyx, content1)
         self.fresh_cythonize(a_pyx)
         self.fresh_cythonize(a_pyx)
         self.assertEqual(1, len(self.cache_files('a.c*')))
-        with open(a_c) as f:
-            a_contents1 = f.read()
+        a_contents1 = self.relative_lines_from_file(
+            a_c, module_line_1, 0, 7)
         os.unlink(a_c)
 
+        self.assertIn(definition_1, a_contents1, INCORRECT)
+
         write_file(a_pyx, content2)
         self.fresh_cythonize(a_pyx)
-        with open(a_c) as f:
-            a_contents2 = f.read()
+        self.assertEqual(2, len(self.cache_files('a.c*')))
+        a_contents2 = self.relative_lines_from_file(
+            a_c, module_line_1, 0, 7)
         os.unlink(a_c)
 
-        self.assertNotEqual(a_contents1, a_contents2, 'C file not changed!')
-        self.assertEqual(2, len(self.cache_files('a.c*')))
+        self.assertNotIn(definition_1, a_contents2, SAME)
+        self.assertIn(definition_2, a_contents2, INCORRECT)
 
         write_file(a_pyx, content1)
         self.fresh_cythonize(a_pyx)
         self.assertEqual(2, len(self.cache_files('a.c*')))
-        with open(a_c) as f:
-            a_contents = f.read()
-        self.assertEqual(
-            a_contents, a_contents1,
-            msg='\n'.join(list(difflib.unified_diff(
-                a_contents.split('\n'), a_contents1.split('\n')))[:10]))
+        a_contents = self.relative_lines_from_file(
+            a_c, module_line_1, 0, 7)
+
+        self.assertIn(definition_1, a_contents, INCORRECT)
+        self.assertNotIn(definition_2, a_contents, INCORRECT)
```

Comment:
.

Reply:
> `difflib` would generally be more convenient to use, but only with `assertEqual`, with `assertNotEqual` I see no way to use the diff.

If `assertNotEqual()` fails, it means that we do not need the (then empty) diff.

I agree that using `difflib` seems nice for unexpected inequality. I also agree that using it in this ad-hoc fashion isn't ideal. There should be a helper function/method for this.

Reply:
> I agree that using `difflib` seems nice for unexpected inequality.

The point is, should we rely on the fact that other tests cover everything else and shouldn't try to see any other unexpected behavior? This looks like additional safety, but it might be superfluous.

Reply:
@scoder resolve this conversation if it is okay not to implement assert with `difflib` in this PR